Update animation_heart and favorite counts when favoriting

Jberczel opened the pull request:
When favoriting an asset, the favorite count does not update until you refresh the page.

In addition, if the asset/mp3 is displayed in multiple sections: "latest uploads", "recently favorited", "latest comments", the other instances of that asset do not get updated (heart svg does not update).

For the favorite counts, I added a turbo_stream update action to update count without having to refresh the page.

For the animation_heart state, I updated stimulus controller to broadcast to other instances of the same asset/mp3 to also toggle.

sudara commented:
Hey this is cool! Thank you! Will pull and check it out on monday.

Jberczel commented:
Hi @sudara - Were you able to take a look? Any issue or concern?

sudara commented:
Hey @Jberczel! Yeah, I took a look before and after you pushed the last commit. I actually did have a Q, i'll ask it inline in the code...

Jberczel commented:
@sudara - I didn't see any inlined comment.

sudara commented:
So sorry! I forgot to hit submit on the "comment review"

Jberczel commented on app/javascript/controllers/favorite_controller.js:
```diff
@@ -5,7 +5,29 @@ export default class extends HeartController {
 
+  connect() {
+    // access controller from child element
+    this.element[this.identifier] = this
```
It " hangs a reference to the Stimulus controller instance off the DOM element that has the same name as the controller itself".

Since there are multiple instances of same asset/mp3, this line allows you to access the other controllers to toggle and keep the favorite count in sync.

Technique was taken from this article:
https://dev.to/leastbad/the-best-one-line-stimulus-power-move-2o90

sudara replied:
Thanks for the article! I've used `getControllerForElementAndIdentifier` elsewhere, so was curious how this differed, whether events would be more appropriate, etc.

Jberczel commented on app/javascript/controllers/favorite_controller.js:
```diff
@@ -5,7 +5,29 @@ export default class extends HeartController {
+  }
+
+  broadcast() {
+    const matches = document.querySelectorAll(`[data-favorite-id-value="${this.idValue}"]`)
```
I'm not sure.  I did try to use Stimulus target API, but those are scoped to individual instances of a controller, and I needed to call toggle on the other instances of the favorites_controller.

sudara commented:
I love the idea of having the counts update + the other hearts fire, but I think we should pick a more straightforward pattern to accomplish this vs. working around stimulus, either:

1. the hearts count / toggle stream from the backend on update
2. custom events update the relevant controllers ([example](https://fullstackheroes.com/tutorials/stimulus/create-custom-events/))

I'll close this PR for now. If you want to take a crack at either of the above methods, feel free. If not, you've still motivated me to care about the subject, so, thank you!!

Jberczel commented:
Sounds good. I'll take a look at custom events.

Jberczel commented:
Hi @sudara - I refactored PR to use a more "stimulus-y" approach to get the other hearts to fire.

As you mentioned earlier, I first implemented using custom event.

But after digging into documentation for Stimulus 3.2, I discovered Stimulus Outlets API, which we can use for cross-controller communication as an alternative to dispatching custom events on controller elements.

See:

https://stimulus.hotwired.dev/reference/outlets

I've implemented with both approaches in two commits depending on which pattern you prefer.

Could you re-open PR?

sudara commented:
Howdy!

Oooh outlets could be cool! If you have commits to both, would you mind pushing them to github so I could take a look. Sorry this wasn’t an insta-marge but it introduces a new pattern, so I’d like to get it right!

Whatever is easiest for you, you can point me to your branches on github or open new PR for them and we’ll merge one (doesn’t make sense to reopen this one)
